File: frontend/src/utils/columnMapper.js
export const columnMap = {
  product: [
    "Product",
    "Product Name",
    "Name",
    "Item",
    "Title"
  ],

  category: [
    "Category",
    "Type",
    "Department"
  ],

  brand: [
    "Brand",
    "Company",
    "Manufacturer"
  ],

  price: [
    "Price",
    "Selling Price",
    "Unit Price",
    "Cost",
    "Amount"
  ],

  stock: [
    "Stock",
    "Inventory",
    "Quantity",
    "Qty",
    "Available"
  ],

  sales: [
    "Sales",
    "Units Sold",
    "Sold",
    "Orders"
  ],

  revenue: [
    "Revenue",
    "Income",
    "Turnover"
  ],

  competitorPrice: [
    "Competitor Price",
    "Market Price",
    "Competitor"
  ],

  month: [
    "Month",
    "Date",
    "Order Month"
  ]
};

export function getValue(row, names) {
  const keys = Object.keys(row);

  for (const key of keys) {
    const normalizedKey = key
      .toLowerCase()
      .replace(/[_\-\s]/g, "");

    for (const name of names) {
      const normalizedName = name
        .toLowerCase()
        .replace(/[_\-\s]/g, "");

      if (normalizedKey === normalizedName) {
        return row[key];
      }
    }
  }

  return "";
}
export function normalizeDataset(data) {
  return data.map((row, index) => ({
    id: index + 1,

    product: getValue(row, columnMap.product),

    category: getValue(row, columnMap.category),

    brand: getValue(row, columnMap.brand),

    price: parseFloat(
      String(getValue(row, columnMap.price))
        .replace(/[$,]/g, "")
    ) || 0,

    stock: parseInt(
      getValue(row, columnMap.stock)
    ) || 0,

    sales: parseInt(
      getValue(row, columnMap.sales)
    ) || 0,

    revenue: parseFloat(
      String(getValue(row, columnMap.revenue))
        .replace(/[$,]/g, "")
    ) || 0,

    competitorPrice: parseFloat(
      String(getValue(row, columnMap.competitorPrice))
        .replace(/[$,]/g, "")
    ) || 0,

    month: getValue(row, columnMap.month),

    raw: row,
  }));
}

File: backend/utils/column_mapper.py
COLUMN_MAP = {

    "product": [
        "product",
        "product_name",
        "product name",
        "name",
        "item",
        "title",
        "product_title"
    ],

    "category": [
        "category",
        "product_category",
        "product category",
        "type",
        "department",
        "product_category_name"
    ],

    "brand": [
        "brand",
        "company",
        "manufacturer",
        "vendor"
    ],

    "price": [
        "price",
        "selling price",
        "selling_price",
        "amount",
        "cost",
        "unit_price",
        "sale_price"
    ],

    "stock": [
        "stock",
        "inventory",
        "qty",
        "quantity",
        "available",
        "inventory_level"
    ],

    "sales": [
        "sales",
        "sold",
        "units sold",
        "orders",
        "order_count"
    ],

    "revenue": [
        "revenue",
        "income",
        "turnover",
        "total_revenue"
    ],

    "competitorPrice": [
        "competitorprice",
        "competitor price",
        "competitor_price",
        "marketprice",
        "market price"
    ],

    "month": [
        "month",
        "date",
        "order month",
        "order_month"
    ]

}


def normalize_dataframe(df):

    rename_dict = {}

    for column in df.columns:

        cleaned = (
            column
            .strip()
            .lower()
            .replace("_", " ")
        )

        mapped = None

        for standard, aliases in COLUMN_MAP.items():

            if cleaned in aliases:

                mapped = standard

                break

        if mapped:

            rename_dict[column] = mapped

        else:

            rename_dict[column] = (
                column
                .strip()
                .replace(" ", "_")
            )

    df = df.rename(columns=rename_dict)

    return df